### PR TITLE
(PUP-7326) Handle unresolvable SIDs in Windows groups

### PR DESCRIPTION
**Pull request by Iristyle**
PR #6647 had a single test failure, but was otherwise ready for merge (but unfortunately accidentally merged). That single fail is addressed in a9d6a83. The additional commits improve some of the messages that appear during error handling.

Comments from original PR #6647:

(PUP-7326) Handle Windows unresolvable SIDs

 - bef10f187f10b5083e73c888d1b4f663ce3a67d9 improved the handling of
   account names within local groups that refer to domain or virtual
   accounts. Prior to that change, Puppet could not manage groups
   with virtual accounts, failed to properly disambiguate accounts
   with the same name that came from different authorities (for
   example local Administrator from NT AUTHORITY vs domain), and
   would produce output from `puppet resource` that didn't include
   the authority.

   Puppet has always emitted errors when managing groups with
   unresolvable SIDs (typically domain accounts that have been deleted
   but that are still present in local groups), but the previous 
   change degraded the behavior of `puppet resource` so that instead
   of producing output like the following when an unresolved SID is
   present:

```puppet
   group { 'Bar':
     ensure  => 'present',
     gid     => 'S-1-5-21-969762586-3490402988-3896818487-1004',
     members => ['SYSTEM', 'Administrator', 'Administrator', 'BM', 'S-1-5-21-3661721861-956923663-2119435483-1116'],
   }
```

   It would instead immediately produce an error like:

```
   Error: Could not run: Failed to call LookupAccountSidW:  No mapping between account names and security IDs was done.
```

   The output is now correct:

```puppet
   group { 'Bar':
     ensure  => 'present',
     gid     => 'S-1-5-21-969762586-3490402988-3896818487-1004',
     members => ['NT AUTHORITY\SYSTEM', 'DFSDEMO\Administrator', 'DFS-DEMO-CLIENT\Administrator', 'DFSDEMO\BM', 'S-1-5-21-3661721861-956923663-2119435483-1116 (unresolvable)'],
   }
```

   The previous change introduced the authority prefixes, but also
   encountered difficulties with unresolvable SIDs, earlier in the
   provider lifecycle.

 - With auth_membership => true, unresolvable SIDs will be removed as
   expected from groups that have them. When auth_membership => false
   unresolvable SIDs will be preserved in local groups.

 - Introduces a breaking internal change by returning Windows Group
   :members property to the Windows ADSI group provider as an array of
   SID::Principal objects rather than an array of string as
   DOMAIN\USER qualified account names.

   While breaking the internal API is generally undesirable, the only
   callers should be Puppet itself.

   This was the simplest path to preserving orphaned / unresolvable
   SIDs that may be part of a local group, but that no longer exist
   in the upstream domain controller, without introducing a number of
   new code paths with special case error-handling.

 - Introduces an additional breaking internal change by changing
   Puppet::Util::Windows::SID::Principal#to_s to return a string in
   DOMAIN\USER format rather than in S-1-XXX style string format.
   This was necessary for `puppet resource` to produce manifests with
   friendly account names rather than SIDs.

   6a3d8215385be5b659db8b4792680273db47da69 initially added this
   method for 3rd party backward compatiblity when the class
   Win32::Security::SID was removed as part of 4.3.2. At the time, all
   internal callsites were modified to use #sid accessor of 
   Puppet::Util::Windows::SID::Principal instead of to_s in the prior
   commit at 42fceb295c4f37ca99b5d0d283b30970745dfadb

 - Note that it is unnecessary to implement the same safeguards
   against the user provider, as it should not be possible to have
   unresolvable SIDs for the groups a user is a member of. Local users
   cannot be added to domain groups.

### CODE REVIEW

File: lib/puppet/util/windows/principal.rb [hunk 1]
```diff
@@ -65,14 +65,14 @@ def self.lookup_account_name(system_name = nil, account_name)
                 last_error = FFI.errno
 
                 if (success == FFI::WIN32_FALSE && last_error != ERROR_INSUFFICIENT_BUFFER)
-                  raise Puppet::Util::Windows::Error.new('Failed to call LookupAccountNameW', last_error)
+                  raise Puppet::Util::Windows::Error.new("Failed to call LookupAccountNameW with account: #{account_name}", last_error)
```
Comment (Iristyle) [inline lib/puppet/util/windows/principal.rb]: Vet whether this is correct for translations after a mergeup...

File: lib/puppet/util/windows/sid.rb [hunk 2]
```diff
@@ -112,12 +112,15 @@ def ads_to_principal(ads_object)
...
-
-      raise if !unresolvable
+      # if the error is not a lookup / mapping problem, immediately re-raise
+      raise if e.code != ERROR_NONE_MAPPED
+
+      # if the Name property isn't formatted like a SID, OR
+      if !valid_sid?(ads_object.Name) ||
+        # if the objectSID doesn't match the Name property, also raise
+        ((converted = octet_string_to_sid_string(ads_object.objectSID)) != ads_object.Name)
+        raise Puppet::Error.new("ads_object Name: #{ads_object.Name} invalid or does not match objectSID: #{ads_object.objectSID} (#{converted})", e)
```
Comment (Iristyle) [inline lib/puppet/util/windows/sid.rb]: This will need to be adjusted for translation on mergeup

**Comment by Iristyle** [comment]
@glennsarti / @michaeltlombardi this is ready for merge based on @glennsarti feedback.

**Comment by glennsarti** [comment]
Appveyor says No.

**Comment by glennsarti** [comment]
Appveyor error is due to;

The parallel spec process which groups specs is throwing an error when it figures out how many spec files exit.
``` bash
An error occurred while loading ./spec/integration/indirector/direct_file_server_spec.rb.
Failure/Error: module_function :octet_string_to_sid_string

NameError:
  undefined method `octet_string_to_sid_string' for module `Puppet::Util::Windows::SID'
# ./lib/puppet/util/windows/sid.rb:235:in `module_function'
# ./lib/puppet/util/windows/sid.rb:235:in `<module:SID>'
# ./lib/puppet/util/windows/sid.rb:4:in `<module:Windows>'
# ./lib/puppet/util/windows/sid.rb:3:in `<top (required)>'
# ./lib/puppet/util/windows.rb:20:in `require'
# ./lib/puppet/util/windows.rb:20:in `<module:Windows>'
# ./lib/puppet/util/windows.rb:1:in `<top (required)>'
# ./lib/puppet/file_system/windows.rb:2:in `require'
# ./lib/puppet/file_system/windows.rb:2:in `<top (required)>'
# ./lib/puppet/file_system.rb:11:in `require'
# ./lib/puppet/file_system.rb:11:in `<module:FileSystem>'
# ./lib/puppet/file_system.rb:1:in `<top (required)>'
```

**Comment by Iristyle** [comment]
Localization follow-on PR at #6695

**Pull request merged**